### DISCUSSION TITLE
Tables not being dropped during testing suite

### DISCUSSION THREAD

**Discussion by alexclavel-ocient** [sqlalchemy/sqlalchemy#9532] [Usage Questions]
Hi SQLAlchemy team,

Background: I am currently writing a sqla connection for my company, Ocient.

I am currently running tests by running `from sqlalchemy.testing.suite import *`

However, after each test, the tables are not being dropped from the database, leading to further tests failing.

Is there a method to drop the tables that I need to implement? Any help would be appreciated.

Thanks,
Alex

**Comment by CaselIT** [comment]
Hi,

There are additional customization points, called provision, in the test that allow customization on how to remove various schema items, but for tables the default configuration should work.

Can you provide some example of test that fail and what log is emitted, maybe running the tests using `--log-info=sqlalchemy.engine`?

**Comment by alexclavel-ocient** [comment]
The issue is on the database side. When running tests, the first test will pass. Afterwards, any test that tries to use the same table name will error out with `A table with that name exists`

Could this be due to us not supporting transactions in Ocient? If that is the case, how can we work around this?

A full log can be found here: https://gist.github.com/alexclavel-ocient/54f254dbf3f911ef6eba36521b6798d3

**Reply by CaselIT** [reply]
> Could this be due to us not supporting transactions in Ocient? If that is the case, how can we work around this?

that's may be a factor in some way, but I don't think it's the main thing, since db that support transactions usually don't fully support transactional ddl. @zzzeek do you think it would cause problem the absence of transactions?

Skimming the logs it seems that there are various different things going on there:
- TypeError: %b requires a bytes-like object, or an object that implements __bytes__, not 'int'
- a statement like `SELECT 1 AS has_table FROM None.some_table` that seems due to a mis compilation of a schema
- `(CREATE INDEX currently only supported on one column. Passed in 2 cols.` I don't know if sqlalchemy have a requirement for tests that use indexes with more than one columns. That's an unusual limitation, so those tests may need to be skipped.

I don't know if they interfere with the teardown of the tables.

I would start by trying to get rid of all other issues, by either fixing the dialect of skipping the test (see an example here of how to skip particular tests on the dialect https://github.com/gordthompson/sqlalchemy-access/blob/master/test/test_suite.py)

**Reply by zzzeek** [reply]
teardown of tables can fall apart if the setup for some tests are failing, or if some tests themselves have really big issues.   

to debug this kind of thing you have to run the suite with -x to find the first test that has a problem, then run just that test by itself and get it to setup, test, and teardown perfectly.

**Reply by zzzeek** [reply]
like this error:

pyocient.Error: State: 42502 Code: -702 Reason: The user does not have the authority to modify the table (CREATE INDEX currently only supported on one column. Passed in 2 cols.)


if the database in question only supports indexes with one column, I would disable "indexes" in the requirements for now, get the suite running, then see if we have index tests that are only using indexes with just one column in there, we probably don't have that right now.

**Reply by alexclavel-ocient** [reply]
Hi @zzzeek, when passing str objects to sqlalchemy, they are automatically lowercased and are not quoted , any Idea what I'm missing for this?

example:
`stmt=insert(my_table).values{'name'='Alex', 'id' = 12)` yields:
`INSERT INTO NAMES (name, id) VALUES (alex, 12)`
Expected:
`INSERT INTO NAMES (name, id) VALUES ('Alex', 12)`

**Reply by zzzeek** [reply]
SQLAlchemy has a casing convention that when you indicate table/column names using an all-lowercase name, the name is considered to be case insensitive if other directives are not passed along with it.  for databases that store / report case insensitive names as UPPERCASE (like Oracle, Firebird) there is a dialect-level attribute [requires_name_normalize](https://github.com/sqlalchemy/sqlalchemy/blob/41e4fdb08722ec9843663500c6121e18c1bb0016/lib/sqlalchemy/engine/default.py#L243-L248) which will have it do this conversion to UPPERCASE when it communicates with the schema tables.

**Reply by CaselIT** [reply]
In all cases the values are never changed, so in any dialect if you pass `Alex` as value sqlalchemy will keep `Alex`.

What mike mentioned is valid only for identifiers (like column name, table names, etc)

**Reply by alexclavel-ocient** [reply]
@CaselIT  that's not what I'm seeing though. The str objects are being modified.

**Reply by lelit** [reply]
Uhm, can you write here the exact code that causes that? The one above, `insert(my_table).values{'name'='Alex', 'id' = 12)` isn't even valid in Python AFAICT.

**Reply by alexclavel-ocient** [reply]
Apologies, The code is `stmt=insert(students).values({'name':'Alex', 'id':12})`
In the database, the query executed is `insert into students (id, name) values (12, alex)`

**Reply by lelit** [reply]
That's still very strange... please copy&paste the exact traceback/log (assuming you have `echo=True` on your connection), and possibly the definition of your `students` table.

**Reply by alexclavel-ocient** [reply]
Log: https://gist.github.com/alexclavel-ocient/4ea73728f6f729b02108dd911c0d7c74
Ignore the erroring out, that is expected, I am using the base compiler visit_insert for testing
The query passed to the database is still `insert into students (id, name) values (12, alex)`

**Reply by CaselIT** [reply]
From the gist:
> 2023-03-30 18:15:06,866 INFO sqlalchemy.engine.Engine b'INSERT INTO students (id, name) VALUES (%(id_m0)s, %(name_m0)s)'
2023-03-30 18:15:06,866 INFO sqlalchemy.engine.Engine [dialect ocient+ocient does not support caching 0.00057s] {b'id_m0': 12, b'name_m0': 'Alex'}

> [SQL: b'INSERT INTO students (id, name) VALUES (%(id_m0)s, %(name_m0)s)']
[parameters: {b'id_m0': 12, b'name_m0': 'Alex'}]

from the log it seems to me that everything is correctly passed by sqlalchemy to the driver

**Reply by alexclavel-ocient** [reply]
Looks like it to me too, however, the server still recieves the lowercased values

**Reply by lelit** [reply]
How can you tell? The gist shows a **syntax error**, followed by a `ROLLBACK`: the syntax error is related to the `INSERT`, and the arguments it receives clearly show the right case, so what do you mean with _"the server still recieves the lowercased values"_?
Just to be on the same page, with _"I am currently writing a sqla connection"_ you mean you are writing a database driver, right?
